sql/connect/common/src/main/protobuf/spark/connect/expressions.proto

```diff
@@ -215,13 +215,19 @@ message Expression {
     }
 
     message Array {
-      DataType element_type = 1;
+      // (Optional) The element type of the array. Only need to set this when the elements are
+      // empty, since spark 4.1+ supports inferring the element type from the elements.
+      optional DataType element_type = 1;
       repeated Literal elements = 2;
```

Comment:
I am not sure whether it is worthwhile to just optimize the `element_type`.
For large arrays of primitive types, e.g. large dense matrix for ML, we introduced `SpecializedArray`.

Reply:
For a `Array[Array[Int]]` case, how to infer the nullability ?

Reply:
The null literal has a data type according to https://github.com/apache/spark/blob/master/sql/connect/common/src/main/protobuf/spark/connect/expressions.proto#L169

Reply:
@zhengruifeng This optimization helps both array and map